src/lib/merkle_ledger/database.ml

```diff
@@ -267,89 +267,79 @@ module Make (Inputs : Inputs_intf) :
-    let last_location () =
-      Location.build_generic (Bigstring.of_string "last_token")
-
-    (* Update the highest token. *)
-    let update_last_location mdb tid =
-      let location = last_location () in
-      let should_update =
-        match get_bin mdb location Token_id.Stable.Latest.bin_read_t with
-        | Some prev_tid ->
-            prev_tid < tid
-        | None ->
-            true
-      in
-      if should_update then
-        set_bin mdb location Token_id.Stable.Latest.bin_size_t
-          Token_id.Stable.Latest.bin_write_t tid
-
-    let last_token mdb =
-      Option.value ~default:Token_id.default
-      @@ get_bin mdb (last_location ()) Token_id.Stable.Latest.bin_read_t
-
-    let get mdb token_id =
-      get_bin mdb (build_location token_id) Key.Stable.Latest.bin_read_t
-
-    let _set mdb token_id public_key =
-      update_last_location mdb token_id ;
-      set_bin mdb (build_location token_id) Key.Stable.Latest.bin_size_t
-        Key.Stable.Latest.bin_write_t public_key
-
-    let _delete mdb token_id = delete_raw mdb (build_location token_id)
-
-    let foldi t ~init ~f =
-      let last_token = last_token t in
-      (* This assumes that token owners are dense in the interval
-         [default+1, last_token]. Otherwise, we should track an indirection,
-         like we do with accounts.
-      *)
-      let rec go acc tid =
-        let tid = Token_id.next tid in
-        if Token_id.(tid <= last_token) then
-          let acc =
-            Option.fold ~init:acc
-              ~f:(fun acc pk -> f ~key:tid ~data:pk acc)
-              (get t tid)
-          in
-          go acc tid
-      in
-      go init Token_id.default
-
-    let _iteri t ~f = foldi t ~init:() ~f:(fun ~key ~data () -> f ~key ~data)
-
-    let get_all t =
-      let last_token = last_token t in
-      (* This assumes that token owners are dense in the interval
-         [default+1, last_token]. Otherwise, we should track an indirection,
-         like we do with accounts.
-      *)
-      let rec go tid =
-        let tid = Token_id.next tid in
-        if Token_id.(tid <= last_token) then
-          match get t tid with
-          | Some pk ->
-              (tid, pk) :: go tid
-          | None ->
-              go tid
-        else []
-      in
-      go Token_id.default
-  end
-
   (** The tokens associated with each public key.
 
       These are represented as a [Token_id.Set.t], which is represented by an
       ordered list.
   *)
   module Tokens = struct
+    let next_available_key =
+      Memo.unit (fun () ->
+          Location.build_generic (Bigstring.of_string "next_available_token")
+      )
+
+    let next_available mdb =
+      Option.value
+        ~default:Token_id.(next default)
+        (get_bin mdb (next_available_key ()) Token_id.Stable.Latest.bin_read_t)
+
+    let next_available_kv ~ledger_depth tid =
+      let token_buf =
+        Bigstring.create (Token_id.Stable.Latest.bin_size_t tid)
```

Comment:
is it clearer to use `Bin_prot.Common.create_buf`?

Reply:
That is nicer, yea. Updated.